src/nupic/research/connections.py

```diff
@@ -243,23 +249,67 @@ def updateSynapsePermanence(self, synapse, permanence):
     self._synapsesForPresynapticCell[newData.presynapticCell][synapse] = newData
 
 
-  def numSegments(self):
-    """
-    Returns the number of segments.
+  def computeActivity(self, activeInput, activePermanenceThreshold,
+                      activeSynapseThreshold, matchingPermananceThreshold,
+                      matchingSynapseThreshold):
+    """ Computes active and matching segments given the current active input.
+
+    @param activeInput                 (set)   currently active cells
+    @param activePermanenceThreshold   (float) permanence threshold for a
+                                               synapse to be considered active
+    @param activeSynapseThreshold      (int)   number of synapses needed for a
+                                               segment to be considered active
+    @param matchingPermananceThreshold (float) permanence threshold for a
+                                               synapse to be considered matching
+    @param matchingSynapseThreshold    (int)   number of synapses needed for a
+                                               segment to be considered matching
+
+    @return (tuple) Contains:
+                      `activeSegments`         (list),
+                      `matchingSegments`        (list),
+
+    Notes:
+      activeSegments and matchingSegments are sorted by the cell they are on.
     """
+
+    numActiveSynapsesForSegment = [0] * self._nextSegmentIdx
+    numMatchingSynapsesForSegment = [0] * self._nextSegmentIdx
+
+    for cell in activeInput:
+      for synapseData in self.synapsesForPresynapticCell(cell).values():
+        segment = synapseData.segment
+        permanence = synapseData.permanence
+        if permanence >= matchingPermananceThreshold:
+          numMatchingSynapsesForSegment[segment] += 1
+          if synapseData.permanence >= activePermanenceThreshold:
+            numActiveSynapsesForSegment[segment] += 1
+
+    activeSegments = []
+    matchingSegments = []
+    for i in xrange(self._nextSegmentIdx):
+      if numActiveSynapsesForSegment[i] >= activeSynapseThreshold:
+        activeSegments.append(i)
+
+    for i in xrange(self._nextSegmentIdx):
+      if numMatchingSynapsesForSegment[i] >= matchingSynapseThreshold:
+        matchingSegments.append(i)
+
+    return (sorted(activeSegments, cmp=self.segmentCmp),
```

Comment:
Using lambda functions is probably easier here since it's such a simple comparison and used only in these two places:

```
segmentCmp = lambda a, b: self._segments[a] - self._segments[b]
return (sorted(activeSegments, cmp=segmentCmp),
        sorted(activeSegments, cmp=segmentCmp))
```